fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,229 @@
+        """
+        assert isinstance(self.temperature, float)
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits = posi.reshape(-1, 1)
+        assert len(local_features) == len(global_features)
+        nega = self.cos_sim(local_features, global_features)
+        logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temperature
+        labels = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def get_perFCL_loss(
+        self,
+        local_features: torch.Tensor,
+        old_local_features: torch.Tensor,
+        global_features: torch.Tensor,
+        old_global_features: torch.Tensor,
+        aggregated_global_features: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        PerFCL loss consists of two contrastive losses.
+        First one airms to enhance the similarity between the current global features and aggregated global feature
+        as positive pairs while reducing the similarity between the current global features and old global
+        features as negative pairs.
+        Second one airms to enhance the similarity between the current local features and old local feature
+        as positive pairs while reducing the similarity between the current local features and aggregated global
+        features as negative pairs.
+        """
+
+        assert isinstance(self.temperature, float)
+        assert len(global_features) == len(aggregated_global_features)
+        posi = self.cos_sim(global_features, aggregated_global_features)
+        logits_max = posi.reshape(-1, 1)
+        assert len(global_features) == len(old_global_features)
+        nega = self.cos_sim(global_features, old_global_features)
+        logits_max = torch.cat((logits_max, nega.reshape(-1, 1)), dim=1)
+        logits_max /= self.temperature
+        labels_max = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits_min = posi.reshape(-1, 1)
+        assert len(local_features) == len(aggregated_global_features)
+        nega = self.cos_sim(local_features, aggregated_global_features)
+        logits_min = torch.cat((logits_min, nega.reshape(-1, 1)), dim=1)
+        logits_min /= self.temperature
+        labels_min = torch.zeros(local_features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits_min, labels_min), self.ce_criterion(logits_max, labels_max)
+
+    def compute_loss(self, preds: Dict[str, torch.Tensor], target: torch.Tensor) -> Losses:
+        if self.old_global_module is None:
+            return super().compute_loss(preds, target)
+        loss = self.criterion(preds["prediction"], target)
+
+        # Optimal cos_sim_loss_weight is 100.0
```

Comment:
Maybe mention where this optimal value is derived from? I imagine the optimal value will differ depending on the downstream application. Is this a good default?

Reply:
Yeah honestly I just put it here based on my observations in fedisic, but I am not sure about other datasets.